File: Shooter/Flame.m
//
//  Flame.m
//  Shooter
//
//  Created by Sven A. Schmidt on 06/11/2013.
//  Copyright (c) 2013 feinstruktur. All rights reserved.
//

#import "Flame.h"

#import "Collider.h"
#import "Constants.h"
#import "FloorSegment.h"
#import "Snowflake.h"
#import "Shooter-Swift.h"


@implementation Flame


+ (instancetype)flameAtPosition:(CGPoint)position
{
    // we need to map SKEmitterNode to Flame before unarchiving
    [NSKeyedUnarchiver setClass:self forClassName:@"SKEmitterNode"];
    Flame *flame = (Flame *)[Flame emitterNodeWithParticleFileNamed:@"flame"];

    flame.position = position;
    flame.name = @"Flame";
    
    flame.physicsBody = [SKPhysicsBody bodyWithRectangleOfSize:CGSizeMake(24, 110)];
    flame.physicsBody.dynamic = NO;
    flame.physicsBody.categoryBitMask = FlameCategory;

    return flame;
}


- (void)collideWith:(SKPhysicsBody *)body
{
    if ([body.node isKindOfClass:[Snowflake class]]) {
        [Collider collideSnowflake:(Snowflake *)body.node withFlame:self];
    }
}


- (CGFloat)bottomEdgeY
{
    // the emitter sits above the base, this is just a number that works well as the "touch point" for the snow
    return self.frame.origin.y + 40;
}


@end

File: Shooter/Flame.swift
//
//  Flame.swift
//  Shooter
//
//  Created by Sven A. Schmidt on 18/06/2014.
//  Copyright (c) 2014 feinstruktur. All rights reserved.
//

import Foundation
import SpriteKit


class Flame: SKEmitterNode, CollisionHandling {
    var bottomEdgeY: CGFloat {
        get {
            return self.frame.origin.y + 40
        }
    }
    
    convenience init(position: CGPoint) {
        self.init()
        
        self.particleTexture = SKTexture(imageNamed: "spark.png")
        self.particleBirthRate = 500
        self.particleLifetime = 0.7
        self.particlePositionRange = CGVectorMake(10, 3)
        self.emissionAngle = CGFloat(M_PI_2)
        self.emissionAngleRange = 5 * CGFloat(M_PI)/180
        self.particleSpeed = 80
        self.particleSpeedRange = 20
        self.particleAlpha = 0.8
        self.particleColorAlphaRange = 0.2
        self.particleColorAlphaSpeed = -0.4
        self.particleScale = 0.2
        self.particleScaleSpeed = -0.1
        self.particleColorBlendFactor = 1
        self.particleColorBlendFactorSpeed = 1
        self.particleColor = UIColor(red: 78.0/255, green: 33.0/255, blue: 6.0/255, alpha: 1)
        self.particleBlendMode = .Add
        
        self.position = position
        self.name = "Flame"
        
        self.physicsBody = SKPhysicsBody(rectangleOfSize: CGSize(width: 24, height: 110))
        self.physicsBody?.dynamic = false
        self.physicsBody?.categoryBitMask = ObjectCategory.Flame.toBitMask()
    }
    
    func collide(body: SKPhysicsBody) {
        if let flake = body.node as? Snowflake {
            Collider.collide(flake: flake, flame: self)
        }
    }
}
